### PR TITLE
VIDEO-7714 - Convert createLocalTracks

### PR DESCRIPTION
**Pull request by PikaJoyce**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details
Conversion of `createLocalTracks` to TypeScript. 

NOTE:
@manjeshbhargav @charliesantos @makarandp0 currently tests are not yet working but we'll get to fixing that as soon as the actual JS to TS conversion looks good to go! 

### Description

JIRA Ticket: [VIDEO-7714](https://issues.corp.twilio.com/browse/VIDEO-7714)

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [x] Updated affected documentation
* [x] Verified locally with `npm test`
* [ ] Manually sanity tested running locally
* [ ] Included screenshot as PR comment (if needed)
* [ ] Ready for review

### CODE REVIEW

File: lib/createlocaltracks.ts [hunk 1]
```diff
@@ -181,9 +190,9 @@ function createLocalTracks(options) {
  *   levels.
  * @property {string} [loggerName='twilio-video'] - The name of the logger. Use this name when accessing the logger used by the SDK.
  *   See [examples](module-twilio-video.html#.connect) for details.
+ * @property {Array<LocalTrack|MediaStreamTrack>} [tracks] - The {@link LocalTrack}s or MediaStreamTracks which is used to construct the LocalTrack.
+ *   These tracks can be obtained by constructing them from the MediaStream.
```
Comment (manjeshbhargav) [inline lib/createlocaltracks.ts:194]: This property is not part of the public API.
```suggestion
```

File: tsdef/types.d.ts [hunk 2]
```diff
@@ -261,3 +264,9 @@ export class StatsReport {
 export interface CancelablePromise<T> extends Promise<T> {
   cancel: () => void;
 }
+
+export type extraLocalTrackOption = CreateLocalTrackOptions & { isCreatedByCreateLocalTracks?: boolean };
```
Comment (manjeshbhargav) [inline tsdef/types.d.ts:268]: This is not a public type. Please move it to `createLocalTracks.ts`.

File: tsdef/types.d.ts [hunk 3]
```diff
@@ -261,3 +264,9 @@ export class StatsReport {
...
+
+export class TwilioError extends Error {
+  code: number;
+}
```
Comment (manjeshbhargav) [inline tsdef/types.d.ts:272]: Not required, since `createLocalTracks()` does not throw a TwilioError.
```suggestion
```

**Review by manjeshbhargav — CHANGES_REQUESTED** [review]

File: lib/createlocaltracks.ts [hunk 4]
```diff
@@ -91,77 +93,84 @@ function createLocalTracks(options) {
...
-    log.debug('LocalTracks:', options.tracks);
-    return Promise.resolve(options.tracks);
+    log.debug('LocalTracks:', config.tracks);
+    return Promise.resolve(config.tracks);
   }
 
-  const extraLocalTrackOptions = {
-    audio: options.audio && options.audio.name
-      ? { name: options.audio.name }
+  const extraLocalTrackOptions: { audio: extraLocalTrackOption; video: extraLocalTrackOption; } = {
```
[the remark this answers is not in the document]
Comment (PikaJoyce) [reply lib/createlocaltracks.ts:124]: @manjeshbhargav I've put it in `tsdef/types.d.ts`.

File: lib/createlocaltracks.ts [hunk 5]
```diff
@@ -1,8 +1,9 @@
 'use strict';
 
-const asLocalTrack = require('./util').asLocalTrack;
-const buildLogLevels = require('./util').buildLogLevels;
-const getUserMedia = require('@twilio/webrtc').getUserMedia;
+import { CreateLocalTrackOptions, CreateLocalTracksOptions, LocalTrack } from '../tsdef/types';
+
+const { asLocalTrack, buildLogLevels } = require('./util');
+const { getUserMedia } = require('@twilio/webrtc');
```
Comment (charliesantos) [inline lib/createlocaltracks.ts:6]: Please add `MediaStreamTrack` and delete line 14

File: lib/createlocaltracks.ts [hunk 6]
```diff
@@ -91,80 +96,89 @@ function createLocalTracks(options) {
...
-    log.debug('LocalTracks:', options.tracks);
-    return Promise.resolve(options.tracks);
+    log.debug('LocalTracks:', fullOptions.tracks);
+    return fullOptions.tracks;
   }
 
-  const extraLocalTrackOptions = {
-    audio: options.audio && options.audio.name
-      ? { name: options.audio.name }
+  const extraLocalTrackOptions: { audio: ExtraLocalTrackOption; video: ExtraLocalTrackOption; } = {
```
Comment (charliesantos) [inline lib/createlocaltracks.ts:127]: Instead of defining the type inline, please create `ExtraLocalTrackOptions` type above (line 25)
```ts
type ExtraLocalTrackOptions = { audio: ExtraLocalTrackOption; video: ExtraLocalTrackOption; };
```

Then in here, you can do
```ts
const extraLocalTrackOptions: ExtraLocalTrackOptions = {...
```

File: package.json [hunk 7]
```diff
@@ -95,11 +95,11 @@
   "types": "./tsdef/index.d.ts",
   "scripts": {
     "lint:js": "eslint ./lib ./test/*.js ./docker/**/*.js ./test/framework/*.js ./test/lib/*.js ./test/integration/** ./test/unit/** ",
-    "lint:ts": "eslint ./tsdef/*.ts ./lib/**/*.ts",
+    "lint:ts": "eslint ./tsdef/*.ts ./lib/**/*.ts ./lib/*.ts",
     "lint": "npm-run-all lint:js lint:ts",
     "printVersion": "node --version && npm --version",
     "test:unit": "npm-run-all printVersion build:es5 test:unit:js test:unit:ts",
-    "test:unit:js": "nyc --report-dir=./coverage/js --extension=.js --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha ./test/unit/index.js",
+    "test:unit:js": "nyc --report-dir=./coverage/js --extension=.js --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha -r ts-node/register ./test/unit/index.js",
```
[the remark this answers is not in the document]
Comment (PikaJoyce) [reply package.json:102]: Great question! Yes, I've tried using `require` instead but since we look at the non-compiled `createLocalTracks` file to do our unit tests, we need `ts-node/register`

[the remark this answers is not in the document]
Comment (charliesantos) [reply package.json:102]: createLocalTracks's unit tests are under `test:unit:ts` script correct? So we don't need `ts-node/register` in `test:unit:js`?

[the remark this answers is not in the document]
Comment (PikaJoyce) [reply package.json:102]: @charliesantos they're not since `createLocalTrack`'s tests are still a JS file.

[the remark this answers is not in the document]
Comment (charliesantos) [reply package.json:102]: Ah that's true. In that case, we should probably combine `test:unit:js` and `test:unit:ts` since these two are running the same commands. Also, I think next step would be to convert the unit test file to TS (after this PR is approved. It can be another jira). We should do that before converting another public API.

[the remark this answers is not in the document]
Comment (PikaJoyce) [reply package.json:102]: I'll create another ticket to convert the unit test file and pull it into this sprint then!

File: lib/createlocaltracks.ts [hunk 8]
```diff
@@ -124,7 +123,9 @@ export async function createLocalTracks(options?: CreateLocalTracksOptions): Pro
     return fullOptions.tracks;
   }
 
-  const extraLocalTrackOptions: { audio: ExtraLocalTrackOption; video: ExtraLocalTrackOption; } = {
+  type ExtraLocalTrackOptions = { audio: ExtraLocalTrackOption; video: ExtraLocalTrackOption; };
```
Comment (charliesantos) [inline lib/createlocaltracks.ts:126]: We should put this at the top (line 25)

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (charliesantos) [reply package.json:102]: Now that you converted the unit tests for createlocaltracks, can we now remove ts-node/register here?

**Review by charliesantos — CHANGES_REQUESTED** [review]
Left a few more comments. And as a reminder, please don't forget to add a createLocalTracks test in tsdef/twilio-video-tests.ts.

File: lib/connect.js [hunk 9]
```diff
@@ -1,9 +1,10 @@
 'use strict';
 
+const { createLocalTracks } = require('./createlocaltracks');
```
Comment (makarandp0) [inline lib/connect.js:3]: you cannot include `ts` files from js. remove this line.

File: lib/createlocaltrack.js [hunk 10]
```diff
@@ -12,7 +11,9 @@ const { DEFAULT_LOG_LEVEL, DEFAULT_LOGGER_NAME } = require('./util/constants');
  */
 function createLocalTrack(kind, options) {
   options = Object.assign({
-    createLocalTracks: defaultCreateLocalTracks,
+    // TODO(joma): remove suppresion once converted to TypeScript.
+    // eslint-disable-next-line no-use-before-define
+    createLocalTracks,
```
Comment (makarandp0) [inline lib/createlocaltrack.js:16]:
```suggestion
```

File: lib/connect.js (hunk 9 above)
Comment (makarandp0) [inline lib/connect.js:3]: you cannot include `ts` files from `js`. remove this line.
```suggestion

```

File: lib/index.ts [hunk 11]
```diff
@@ -19,19 +20,25 @@ const internals = {
   LocalVideoTrack: require('./media/track/es5').LocalVideoTrack
 };
 
+interface CreateLocalTrackOptionsInternal extends CreateLocalTrackOptions {
+  createLocalTracks: typeof createLocalTracks
+}
+
 function connect(token: string, options?: ConnectOptions): Promise<Room> {
+  const internalOptions = options as ConnectOptionsInternal;
+  internalOptions.createLocalTracks = createLocalTracks;
```
Comment (makarandp0) [inline lib/index.ts:29]: options is optional...so we cannot dereference it directly.
```suggestion
  const internalOptions = {
     createLocalTracks,
     ...options
  }
```

File: lib/index.ts [hunk 12]
```diff
@@ -19,19 +20,25 @@ const internals = {
...
   return internals.connect(token, options);
 }
 
 function createLocalAudioTrack(options?: CreateLocalTrackOptions): Promise<LocalAudioTrackType> {
+  const internalOptions = options as CreateLocalTrackOptionsInternal;
+  internalOptions.createLocalTracks = createLocalTracks;
```
Comment (makarandp0) [inline lib/index.ts:35]: same here.
```suggestion
  const internalOptions = {
     createLocalTracks,
     ...options
  };
  ```

File: lib/index.ts [hunk 13]
```diff
@@ -19,19 +20,25 @@ const internals = {
...
   return internals.createLocalAudioTrack(options);
 }
 
-function createLocalTracks(options?: CreateLocalTracksOptions): Promise<LocalTrack[]> {
-  return internals.createLocalTracks(options);
-}
-
 function createLocalVideoTrack(options?: CreateLocalTrackOptions): Promise<LocalVideoTrackType> {
+  const internalOptions = options as CreateLocalTrackOptionsInternal;
+  internalOptions.createLocalTracks = createLocalTracks;
```
Comment (makarandp0) [inline lib/index.ts:41]: and again
```suggestion
  const internalOptions = {
     createLocalTracks,
     ...options
  }
```

**Review by makarandp0 — CHANGES_REQUESTED** [review]
few corrections to you recent commit. I have not looked at whole PR though

File: lib/index.ts (hunk 11 above)
Comment (PikaJoyce) [reply lib/index.ts:29]: This makes sense. But this would also mean we would no longer need `ConnectOptionsInternal`?

Comment (makarandp0) [reply lib/index.ts]: yes, `ConnectOptionsInternal` may not be necessary since connect is not yet converted to typescript.

File: lib/createlocaltracks.ts [hunk 14]
```diff
@@ -77,11 +81,11 @@ let createLocalTrackCalls = 0;
  * });
  *
  */
-function createLocalTracks(options) {
+export async function createLocalTracks(options?: CreateLocalTracksOptions): Promise<LocalTrack[]> {
   const isAudioVideoAbsent =
     !(options && ('audio' in options || 'video' in options));
 
-  options = Object.assign({
```
Comment (makarandp0) [inline lib/createlocaltracks.ts:84]: just curious - why was this change necessary? Did typescript not allow:   
`options  = Object.assign({...}, options)`

Comment (makarandp0) [reply lib/createlocaltracks.ts:84]: looks like rest of the changes are caused by options => fullOptions. They can be avoided if we can keep the variable same.

**Review by makarandp0 — COMMENTED** [review]
This is looking good. Please update the base branch so that we do not need to review unrelated changes.

File: lib/createlocaltracks.ts (hunk 14 above)
Comment (charliesantos) [reply lib/createlocaltracks.ts:84]: I think it's because `options` is of type `CreateLocalTracksOptions` and it's public. The properties we're merging into fullOptions doesn't exists on `CreateLocalTracksOptions` and will cause errors if we keep the variable name. We're essentially creating a new type here since we're adding more properties. Maybe we should define it? Not sure since it's only use here.

**Review by charliesantos — CHANGES_REQUESTED** [review]
Looking good @PikaJoyce ! I left a few questions. Please update the base branch and I will take a look again.

File: lib/connect.js [hunk 15]
```diff
@@ -235,7 +234,6 @@ function connect(token, options) {
 
   options = Object.assign({
     automaticSubscription: true,
-    createLocalTracks,
```
[the remark this answers is not in the document]
Comment (PikaJoyce) [reply lib/connect.js:238]: It's because we've moved the dependency into the `index.ts` file instead of having `createLocalTracks` here. This was the suggestion @makarandp0 made to get around relying on a ts module from within our js files.

File: lib/index.ts [hunk 16]
```diff
@@ -20,19 +20,27 @@ const internals = {
 };
 
 function connect(token: string, options?: ConnectOptions): Promise<Room> {
-  return internals.connect(token, options);
+  const internalOptions = {
```
Comment (PikaJoyce) [inline lib/index.ts:23]: So we pass in `createLocalTracks` here @charliesantos. Same for `createLocalVideoTrack` and `createLocalAudioTrack`.

File: lib/connect.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/connect.js:238]: thats because `index.ts` now passes createLocalTracks to `connect`

File: lib/createlocaltrack.js [hunk 17]
```diff
@@ -12,7 +11,6 @@ const { DEFAULT_LOG_LEVEL, DEFAULT_LOGGER_NAME } = require('./util/constants');
  */
 function createLocalTrack(kind, options) {
   options = Object.assign({
-    createLocalTracks: defaultCreateLocalTracks,
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/createlocaltrack.js:15]: same, `index.ts` passes this now.

File: package.json [hunk 18]
```diff
@@ -98,9 +98,7 @@
     "lint:ts": "eslint ./tsdef/*.ts ./lib/**/*.ts",
     "lint": "npm-run-all lint:js lint:ts",
     "printVersion": "node --version && npm --version",
-    "test:unit": "npm-run-all printVersion build:es5 test:unit:js test:unit:ts",
-    "test:unit:js": "nyc --report-dir=./coverage/js --extension=.js --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha ./test/unit/index.js",
-    "test:unit:ts": "nyc --report-dir=./coverage/ts --extension=.ts --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha -r ts-node/register ./test/unit/index.ts",
+    "test:unit": "npm-run-all printVersion build:es5 && nyc  --report-dir=./coverage/tests --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha -r ts-node/register ./test/unit/*",
```
Comment (charliesantos) [inline package.json:101]:
```suggestion
    "test:unit": "npm-run-all printVersion build:es5 && nyc --report-dir=./coverage --include=lib/**/* --reporter=html --reporter=lcov --reporter=text mocha -r ts-node/register ./test/unit/*",
```

**Review by charliesantos — CHANGES_REQUESTED** [review]
please add a createLocalTracks test in tsdef/twilio-video-tests.ts.

**Review by charliesantos — APPROVED** [review]
Please don't forget to add a changelog.
